File: go/internal/refs/refs.go
package refs

import (
	"regexp"
	"strconv"
)

// Kind IDs agents already paste: quest=23, note=12, …
var TokenRe = regexp.MustCompile(`\b(note|quest|questline|step|attachment)=(\d+)\b`)

type Ref struct {
	Kind string `json:"kind"`
	ID   int64  `json:"id"`
}

func Parse(text string) []Ref {
	if text == "" {
		return nil
	}
	seen := map[string]struct{}{}
	var out []Ref
	for _, m := range TokenRe.FindAllStringSubmatch(text, -1) {
		id, err := strconv.ParseInt(m[2], 10, 64)
		if err != nil || id <= 0 {
			continue
		}
		key := m[1] + "=" + m[2]
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, Ref{Kind: m[1], ID: id})
	}
	return out
}

func FilterKind(list []Ref, kind string) []Ref {
	var out []Ref
	for _, r := range list {
		if r.Kind == kind {
			out = append(out, r)
		}
	}
	return out
}

File: frontend/src/lib/js/refs.js
/** Tokens agents already paste: note=12, quest=23, … */

export const REF_RE = /\b(note|quest|questline|step|attachment)=(\d+)\b/g

export const REF_KINDS = ['note', 'quest', 'questline', 'step', 'attachment']

/**
 * @param {string} text
 * @returns {{ kind: string, id: number }[]}
 */
export function parseRefs(text) {
  const seen = new Set()
  const out = []
  const src = String(text || '')
  for (const m of src.matchAll(REF_RE)) {
    const kind = m[1]
    const id = Number(m[2])
    const key = `${kind}=${id}`
    if (!id || seen.has(key)) continue
    seen.add(key)
    out.push({ kind, id })
  }
  return out
}

/** @param {string} kind @param {number} id */
export function refHref(kind, id) {
  return `?${kind}=${id}`
}

/** @param {string} href */
export function parseRefHref(href) {
  try {
    const u = new URL(href, 'http://local')
    for (const kind of REF_KINDS) {
      const raw = u.searchParams.get(kind)
      if (!raw) continue
      const id = Number(raw)
      if (Number.isFinite(id) && id > 0) return { kind, id }
    }
  } catch {
    /* ignore */
  }
  return null
}
